File: 0120-triangle/0120-triangle.py
class Solution:
    def solve(self, triangle: List[List[int]], dp: List[int], i: int, j: int) -> int:
        if i == len(triangle):
            return 0
        idx = (i * (i + 1) // 2) + j
        if dp[idx] == None:
            dp[idx] = triangle[i][j] + min(self.solve(triangle, dp, i + 1, j), self.solve(triangle, dp, i + 1, j + 1))
        return dp[idx]
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        n = len(triangle) * (len(triangle) + 1) // 2
        dp = [None for i in range(n)]
        return self.solve(triangle, dp, 0, 0)

File: 0120-triangle/0120-triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        int n = triangle.size();
        vector<int> dp(n * (n + 1) / 2, INT_MAX);
        return solve(triangle, dp, 0, 0);
    }
private:
    int solve(vector<vector<int>>& triangle, vector<int>& dp, int i, int j)
    {
        if(i == triangle.size())
        {
            return 0;
        }
        int idx = (i * (i + 1) / 2) + j;
        if(dp[idx] == INT_MAX)
        {
            dp[idx] = triangle[i][j];
            dp[idx] += min(solve(triangle, dp, i + 1, j), solve(triangle, dp, i + 1, j + 1));
        }
        return dp[idx];
    }
};